plugins/automelt.cpp

```diff
@@ -1,294 +1,705 @@
-#include "uicommon.h"
 
-#include "modules/Gui.h"
+#include "Debug.h"
+#include "LuaTools.h"
+#include "PluginManager.h"
+#include "TileTypes.h"
+
+#include "modules/Buildings.h"
+#include "modules/Maps.h"
+#include "modules/Items.h"
+#include "modules/World.h"
+#include "modules/Designations.h"
+#include "modules/Persistence.h"
+#include "modules/Units.h"
+
+// #include "uicommon.h"
 
 #include "df/world.h"
+#include "df/building.h"
 #include "df/world_raws.h"
 #include "df/building_def.h"
 #include "df/viewscreen_dwarfmodest.h"
 #include "df/building_stockpilest.h"
-#include "modules/Buildings.h"
-#include "modules/Items.h"
 #include "df/plotinfost.h"
-#include "modules/Maps.h"
-#include "modules/World.h"
 #include "df/item_quality.h"
 
+#include <map>
+#include <unordered_map>
+
 using df::building_stockpilest;
+using std::map;
+using std::multimap;
+using std::pair;
+using std::string;
+using std::unordered_map;
+using std::vector;
+
+using namespace DFHack;
+using namespace df::enums;
 
 DFHACK_PLUGIN("automelt");
-#define PLUGIN_VERSION 0.3
 REQUIRE_GLOBAL(gps);
 REQUIRE_GLOBAL(world);
 REQUIRE_GLOBAL(cursor);
 REQUIRE_GLOBAL(plotinfo);
 
-static const string PERSISTENCE_KEY = "automelt/stockpiles";
+namespace DFHack
+{
+    DBG_DECLARE(automelt, status, DebugCategory::LINFO);
+    DBG_DECLARE(automelt, cycle, DebugCategory::LINFO);
+    DBG_DECLARE(automelt, perf, DebugCategory::LTRACE);
+}
+
+static const string CONFIG_KEY = string(plugin_name) + "/config";
+static const string STOCKPILE_CONFIG_KEY_PREFIX = string(plugin_name) + "/stockpile/";
+static PersistentDataItem config;
+
+static vector<PersistentDataItem> watched_stockpiles;
+static unordered_map<int, size_t> watched_stockpiles_indices;
+
 
-static int mark_item(df::item *item, df::item_flags bad_flags, int32_t stockpile_id)
+enum ConfigValues
 {
-    if (item->flags.whole & bad_flags.whole)
-        return 0;
+    CONFIG_IS_ENABLED = 0,
```

Comment:
automelt doesn't have quite the same semantics for enabling as autochop. For autochop, when it's enabled, it starts having an effect on the game. In this case, there is a required manual step for players to take before this plugin does anything.

I recommend a model closer to buildingplan: https://github.com/DFHack/dfhack/pull/2641
where it does not store its enabled state, but rather is enabled unconditionally in the default dfhack.init (but can still be disabled per-session for debugging or preference)

The business logic here in general is all still applicable and correct (and looks good!), but the top level logic should be swapped out for something based on buildingplan.

Reply:
So if I understand correctly, your suggestion is to assume that it's enabled, provided that there are any stockpiles with the monitor/melt condition enabled?
(And in this case, "assume it's enabled" means adding that entry to the `dfhack.tools.init` config)

That makes sense, you're right that additional steps are required for it to do anything. Have to pick piles to enable etc. before it will have an effect on the game and designate items. I can look at making that change as well.

Reply:
That's correct. This plugin should always *load* and *persist* its state, and if it's disabled, the cycle simply won't run.

Reply:
Gotcha. That's what the latest ver. does.